packages/react-components/src/gallery/dominantSpeaker.ts

```diff
@@ -3,59 +3,78 @@
   visibleParticipants: VideoGalleryRemoteParticipant[] = [],
-  MAX_TILES = 4
+  maxTiles = 4
 ): VideoGalleryRemoteParticipant[] => {
   if (!participants) return [];
 
-  // Only use the Max allowed dominant speakers.
-  const dominantSpeakerIds = dominantSpeakers.slice(0, MAX_TILES);
-
   // Don't apply any logic if total number of video streams is less than Max video streams.
-  if (participants.length <= MAX_TILES) return participants;
+  if (participants.length <= maxTiles) return participants;
 
-  // Initialize `visibleParticipants` if it is empty.
-  if (!visibleParticipants.length) {
-    visibleParticipants = participants.slice(0, MAX_TILES);
+  let currentParticipants = visibleParticipants.slice();
+  // Initialize `currentParticipants` if it is empty.
+  if (!currentParticipants.length) {
+    currentParticipants = participants.slice(0, maxTiles);
   }
 
-  const lastVisibleSpeakerIds = visibleParticipants.map((speaker) => speaker.userId);
+  // Only use the Max allowed dominant speakers.
+  const dominantSpeakerIds = Array.from(new Set(dominantSpeakers).values()).slice(0, maxTiles);
+  const lastVisibleSpeakerIds = currentParticipants.map((p) => p.userId);
   const newDominantSpeakerIds = dominantSpeakerIds.filter((id) => !lastVisibleSpeakerIds.includes(id));
 
   // Remove participants that are no longer dominant and replace them with new dominant speakers.
-  lastVisibleSpeakerIds.forEach((id, idx) => {
-    if (!dominantSpeakerIds.includes(id)) {
+  for (let index = 0; index < maxTiles; index++) {
+    const oldUserId = lastVisibleSpeakerIds[index];
+
+    if ((oldUserId && !dominantSpeakerIds.includes(oldUserId)) || oldUserId === undefined) {
       const replacement = newDominantSpeakerIds.shift();
-      if (replacement) {
-        lastVisibleSpeakerIds[idx] = replacement;
-      }
+      if (!replacement) break;
+      lastVisibleSpeakerIds[index] = replacement;
     }
+  }
+
+  // Search optimization
+  const lastVisibleSpeakersMap = {};
+  lastVisibleSpeakerIds.forEach((userId) => {
+    lastVisibleSpeakersMap[userId] = true;
+  });
+
+  const newSpeakers = participants.filter((p) => lastVisibleSpeakersMap[p.userId]);
+  const newSpeakersMap = {}; // search optimization
+  newSpeakers.forEach((p) => {
+    newSpeakersMap[p.userId] = true;
   });
 
   // Sort the new video participants to match the order of last visible participants.
-  const videoParticipantsToRender = participants.filter((p) => lastVisibleSpeakerIds.includes(p.userId));
-  videoParticipantsToRender.sort((a, b) => {
+  newSpeakers.sort((a, b) => {
     return lastVisibleSpeakerIds.indexOf(a.userId) - lastVisibleSpeakerIds.indexOf(b.userId);
   });
 
   // Add additional participants to the final list of visible participants if the list has less than Max visible participants.
-  if (videoParticipantsToRender.length < MAX_TILES) {
-    const diff = MAX_TILES - videoParticipantsToRender.length;
+  if (newSpeakers.length < maxTiles) {
```

Comment:
Might be cleaner to do this operation on the `newParticipantIDs` list above. Right now, some of the inclusion and ordering decision is made on the ids list above, and some of it is made here.

Instead, arrive at the already sorted list of IDs, *then* go find the participants for each of the IDs.

Reply:
I would suggest something slightly different, instead of maintaining a `newSpeakersMap`, have instead:
```
const newVisibleParticipants = participants.filter((p) => lastVisibleSpeakerIds.includes(p.userId))
const newNonVisibleParticipants = participants.filter((p) => !lastVisibleSpeakerIds.includes(p.userId))
// Ensure we have maxTiles number of visible participants
if (newVisibleParticipants.length < maxTiles) {
  newVisibleParticipants.push(...newNonVisibleParticipants.slice(0, maxTiles - newVisibleParticipants.length);
}
```

The inefficiencies should be negligible given these are tiny arrays, and given the complex function I'd put more weight on readability

Reply:
I think you are only thinking about video tiles. But this will also be used to handle audio participants where the limits imposed by maxtiles won't apply. For example, we are talking about 10 visible audio participants in a call with 100+ audio participants.
This operation happening every time someone speaks. It can add up really fast...

Reply:
I have added some optimizations and removed unneeded code based on your comments.
Should be more readable now.